### PR TITLE
feat: add DFlash2 training support

### PR DESCRIPTION
**Pull request by abatilo**
## Summary

- Add DFlash2 configuration, grouped dynamic causal convolution, candidate selection, and a thin wrapper around the shared DFlash training path.
- Export the public `DFlash2DraftModel` configuration and checkpoint schema used by SGLang and vLLM.
- Add fail-fast checks for unsupported configuration values that would change training-to-serving behavior.
- Add the Qwen3-8B recipe and focused tests for configuration, masking, loss, gradients, selector behavior, distributed export, and artifact schema.

## Public inference contract

TorchSpec follows the public DFlash2 architecture and serving behavior described by the [Inco overview](https://inco.ai/blog/dflash2/), [`z-lab/dflash`](https://github.com/z-lab/dflash/blob/07ebd93db9f472af339b644bb70221ad8428328a/dflash/model.py#L478-L547), [SGLang PR #35371](https://github.com/sgl-project/sglang/pull/35371), and [vLLM PR #52816](https://github.com/vllm-project/vllm/pull/52816):

- Apply grouped dynamic causal convolution around attention and the multilayer perceptron.
- Select the strict unary top-K candidates at each position.
- Apply the output multiplier and optional softcap.
- Add the predecessor-conditioned bilinear transition score.
- Use the verified anchor for the first predecessor, followed by the selected candidate from the previous position.
- Export the public configuration and checkpoint parameter names.

## Training objective boundary

The public sources describe the selector architecture and inference computation. They do not publish selector targets, a loss, negative construction, initialization, or a training schedule.

TorchSpec therefore defines a teacher-forced cross-entropy objective. Training uses the gold predecessor and inserts the gold successor when it is outside the unary top-K candidates. This insertion is training-only. Serving retains the strict public top-K set.

This pull request does not claim parity with an unpublished upstream training recipe.

## Validation

- Focused tests compare the grouped convolution and selector behavior with the public implementation and cover configuration, masks, positions, loss, gradients, distributed export, and checkpoint schema.
- Repository pre-commit checks pass.
- [Lint](https://github.com/lightseekorg/TorchSpec/actions/runs/32280874846) and the [CPU-safe test suite](https://github.com/lightseekorg/TorchSpec/actions/runs/32280909638) pass at the current pull-request head. The CPU-safe suite reports 580 passed and 44 skipped.
- The Qwen3-8B recipe was also exercised through training, export, and SGLang serving in a private test environment. Those artifacts are not linked here and do not replace the public repository checks.

## Limitations

- For an artifact with `block_size=B`, configure vLLM with `num_speculative_tokens=B-1`.
- The public sources do not disclose the upstream selector-training recipe or a production acceptance target. The included recipe and private integration exercise are not evidence of generalization or production throughput.
- The current configuration path rejects mixed full/sliding attention layouts and non-default RoPE scaling instead of exporting a model with different behavior.
- SGLang compatibility was exercised against its public DFlash2 implementation. The vLLM integration was source-reviewed but was not runtime-validated for this change.

## Related work

- DFlash: #64
- DSpark: #129
- Shared DFlash and DSpark training path: #134

### CODE REVIEW

File: tools/convert_to_hf.py [hunk 1]
```diff
@@ -485,19 +500,28 @@ def _convert_fsdp_to_hf(
...
+            raise RuntimeError(
+                "DFlash2 checkpoint state does not match the configured model: "
+                f"missing={invalid_missing}, unexpected={sorted(unexpected)}"
+            )
+    else:
+        missing, unexpected = hf_model.load_state_dict(model_state, strict=False)
+        if missing:
+            logger.warning("Missing keys: %s", missing)
+        if unexpected:
+            logger.warning("Unexpected keys: %s", unexpected)
```
Comment (Dogacel) [inline tools/convert_to_hf.py]:
```suggestion
    missing, unexpected = hf_model.load_state_dict(model_state, strict=False)
    allowed_missing = {"embed_tokens.weight"} if target_model_path else set()
    missing = sorted(set(missing) - allowed_missing)
    if invalid_missing:
        logger.warning("Missing keys: %s", missing)
    if unexpected:
        logger.warning("Unexpected keys: %s", unexpected)
```

**Comment by yubofredwang** [comment]
Thanks for the contribution!

Here is some ongoing training with the implementation indicating the convergence
Screenshot 2026-08-20 at 12 27 58 AM

**Review by yubofredwang — APPROVED** [review]

**Comment by yuyangxie96** [comment]
@yubofredwang  Can you successfully reproduce the results of dflash2 using this PR?

**Comment by abatilo** [comment]
@yuyangxie96 I didn't try reproducing the exact open source models but I did try a 6 way comparison to train Qwen3-8b with a regenerated dataset based on these prompts: https://huggingface.co/datasets/lightseekorg/kimi-mtp-dataset 


DFlash, DFlash2, DSpark each with either the default cross entropy loss or D-PACE based loss. DFlash2 + D-PACE won 13 of the 16 categories in my eval set after 1 epoch.

image
